mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -254,16 +247,46 @@ fn try_get_header_value(headers: &Headers, key: &str) -> Option<String> {
     headers.get(key).ok()?.value().clone().into_string().ok()
 }
 
-fn extract_result(response_headers: &Mutex<Option<Headers>>) -> PutObjectResult {
-    let response_headers = response_headers
-        .lock()
-        .expect("must be able to acquire headers lock")
-        .take()
-        .expect("PUT response headers must be available at this point");
-    PutObjectResult {
+fn get_etag(response_headers: &Headers) -> Result<ETag, ParseError> {
+    Ok(response_headers
+        .get(ETAG_HEADER_NAME)?
+        .value()
+        .clone()
+        .into_string()
+        .map_err(ParseError::Invalid)?
+        .into())
+}
+
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum ParseError {
+    #[error("Header response error: {0}")]
+    Header(#[from] HeadersError),
+
+    #[error("Header string was not valid: {0:?}")]
+    Invalid(OsString),
+}
+
+fn extract_result(response_headers: Headers) -> Result<PutObjectResult, S3RequestError> {
+    let etag = get_etag(&response_headers).map_err(|e| S3RequestError::InternalError(Box::new(e)))?;
+
+    Ok(PutObjectResult {
+        etag,
         sse_type: try_get_header_value(&response_headers, SSE_TYPE_HEADER_NAME),
         sse_kms_key_id: try_get_header_value(&response_headers, SSE_KEY_ID_HEADER_NAME),
-    }
+    })
+}
+
+/// Creates `on_headers` callback that will send the response headers to the matching `Receiver`.
+fn response_headers_handler() -> (impl FnMut(&Headers, i32), Receiver<Headers>) {
+    let (response_headers_sender, response_headers) = oneshot::channel();
+    let mut response_headers_sender = Some(response_headers_sender);
```

Comment:
It took me a while to understand why do we need this. Can we add some comments?

As far as I know, `on_headers` is FnMut because meta request can call `on_headers` callback multiple times depending on request type, but for `put_object` and `put_object_single` we know that it only get called at completion. Is that correct?

Reply:
That's correct. Will add comment.